src/main/assisted-installer-controller/assisted_installer_main.go

```diff
@@ -55,38 +62,77 @@ func main() {
 		kc,
 	)
 
+	assistedController.SetReadyState()
+
 	// While adding new routine don't miss to add wg.add(1)
 	// without adding it will panic
 	var wg sync.WaitGroup
-	var wgLogs sync.WaitGroup
 	var status assistedinstallercontroller.ControllerStatus
 
-	ctxApprove, cancelApprove := context.WithCancel(context.Background())
-	go assistedController.ApproveCsrs(ctxApprove, &wg)
+	ctxRoutines, cancelRoutines := context.WithCancel(context.Background())
+	go assistedController.WaitAndUpdateNodesStatus(ctxRoutines, &wg)
 	wg.Add(1)
-	go assistedController.PostInstallConfigs(&wg, &status)
+	go assistedController.PostInstallConfigs(ctxRoutines, &wg, &status)
 	wg.Add(1)
-	go assistedController.UpdateBMHs(&wg)
+	go assistedController.UpdateBMHs(ctxRoutines, &wg)
 	wg.Add(1)
+
+	// No need to cancel with context, will finish quickly
 	go assistedController.HackDNSAddressConflict(&wg)
 	wg.Add(1)
 
-	ctxLogs, cancelLogs := context.WithCancel(context.Background())
-	go assistedController.UploadLogs(ctxLogs, cancelLogs, &wgLogs, &status)
-	wgLogs.Add(1)
+	go assistedController.UploadLogs(ctxRoutines, &wg, &status)
+	wg.Add(1)
 
-	assistedController.SetReadyState()
-	assistedController.WaitAndUpdateNodesStatus(&status)
-	logger.Infof("Sleeping for 10 minutes to give a chance to approve all csrs")
-	time.Sleep(10 * time.Minute)
-	cancelApprove()
+	waitForInstallation(client, logger, &status)
+	// stop all go routines
+	cancelRoutines()
 
 	logger.Infof("Waiting for all go routines to finish")
 	wg.Wait()
-	if !status.HasError() {
-		//with error the logs are canceled within UploadLogs
-		logger.Infof("closing logs...")
-		cancelLogs()
+	logger.Infof("Finished all")
+}
+
+func waitForInstallation(client inventory_client.InventoryClient, log logrus.FieldLogger, status *assistedinstallercontroller.ControllerStatus) {
```

Comment:
This function should be in the controller file for better maintainability

Reply:
don't fully agree. I use exit(0) and i think better to have it in main